packages/dapp/components/lib/contracts-shortcuts.ts

```diff
@@ -1,11 +1,11 @@
-import { BigNumber, ethers } from "ethers";
+import { BigNumber, ethers, Contract } from "ethers";
 import { ERC1155Ubiquity, ERC20 } from "types";
 
 import { performTransaction } from "./utils";
 
 export async function ensureERC20Allowance(
   logName: string,
-  contract: ERC20,
+  contract: ERC20 | Contract,
```

Comment:
Seems like a red flag that you're adding a generic Contract type here if its for ERC20 token contracts only.

Reply:
> Seems like a red flag that you're adding a generic Contract type here if its for ERC20 token contracts only.

Yeah. its' usecase is as follows
```
await ensureERC20Allowance("uCR -> CreditNftManagerFacet", contracts.dollarToken, amount, signer, contracts.creditNftManagerFacet.address);
```
and the type of `contracts.dollarToken` is `Contract | null` when initialzing, please check [this](https://github.com/ubiquity/ubiquity-dollar/blob/000bdb64f645f4338f6d827546a2b5b7b66fff0d/packages/dapp/components/lib/hooks/contracts/use-protocol-contracts.ts#L63)

What's the recommendation? change the dollar token type as `ERC20` in stead of `Contract`?

Reply:
> Seems like a red flag that you're adding a generic Contract type here if its for ERC20 token contracts only.

```
import { Provider } from "@ethersproject/providers";
import { Contract, ethers } from "ethers";
import { ERC1155Ubiquity, ERC20 } from "types";

import latestDeployment from "@ubiquity/contracts/broadcast/05_StakingShare.s.sol/31337/run-latest.json";
import useWeb3 from "../use-web-3";

// contract build artifacts
// separately deployed contracts
import CreditNftArtifact from "@ubiquity/contracts/out/CreditNft.sol/CreditNft.json";
import UbiquityCreditTokenArtifact from "@ubiquity/contracts/out/UbiquityCreditToken.sol/UbiquityCreditToken.json";
import UbiquityDollarTokenArtifact from "@ubiquity/contracts/out/UbiquityDollarToken.sol/UbiquityDollarToken.json";

export type ProtocolContracts = ReturnType<typeof useProtocolContracts> | null;
const useProtocolContracts = async () => {
  // get current web3 provider
  const { provider } = useWeb3();

  // all protocol contracts
  const protocolContracts: {
    // separately deployed contracts (i.e. not part of the diamond)
    creditNft: ERC1155Ubiquity | null,
    creditToken: ERC20 | null,
    dollarToken: ERC20 | null,
    governanceToken: Contract | null,
  } = {
    // separately deployed contracts (i.e. not part of the diamond)
    creditNft: null,
    creditToken: null,
    dollarToken: null,
    governanceToken: null,
  };


  // for all of the deployment transactions
  latestDeployment.transactions.map(tx => {
    if (tx.transactionType === 'CREATE') {
      // find contracts that deployed separately (i.e. not part of the diamond)
      if (tx.contractName === 'CreditNft') {
        protocolContracts.creditNft = new ethers.Contract(tx.contractAddress, CreditNftArtifact.abi, <Provider>provider);
      }
      if (tx.contractName === 'UbiquityCreditToken') {
        protocolContracts.creditToken = new ethers.Contract(tx.contractAddress, UbiquityCreditTokenArtifact.abi, <Provider>provider);
      }
      if (tx.contractName === 'UbiquityDollarToken') {
        protocolContracts.dollarToken = new ethers.Contract(tx.contractAddress, UbiquityDollarTokenArtifact.abi, <Provider>provider);
      }
    }
  });

  return protocolContracts;
};

export default useProtocolContracts;
```

I tried to define the type of `dollarToken` and `creditToken` as `ERC20`, but couldn't initialize it with `new ethers.Contract()`, seems there is no better approach right now

Reply:
> You should be able to use the type that is generated from the contract itself using typechain, not a generic `Contract` type.

Yeah, I understand what you mean, but in this case, I couldn't match the types.

In [use-protocol-contracts.ts](https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/dapp/components/lib/hooks/contracts/use-protocol-contracts.ts), contracts are initialized with `new ethers.Contract(contractAddress, abi, provider)`, so the type should be general `Contract`.
And the types of the contracts in the functions that is using protocol contracts should also `Contract`, and if defined as custom type, will mismatch.

Reply:
the thing here is diamond architecture is typed as `Contract` and the old type is just a simple ERC20, this should be purged, in fact this whole function should be refactored https://github.com/ubiquity/ubiquity-dollar/blob/000bdb64f645f4338f6d827546a2b5b7b66fff0d/packages/dapp/components/lib/contracts-shortcuts.ts#L8